src/algorithms/matching.js

```diff
@@ -0,0 +1,1257 @@
+ * Bipartite graphs can also be matched using the functions present in
+ * :mod:`networkx.algorithms.bipartite.matching`.
+ *
+ * ### References
+ *
+ * [1] "Efficient Algorithms for Finding Maximum Matching in Graphs",
+ *     Zvi Galil, ACM Computing Surveys, 1986.
+ *
+ * @param {Graph} G
+ *   Undirected graph
+ * @param {boolean=} maxcardinality (default=False)
+ *    If maxcardinality is True, compute the maximum-cardinality matching
+ *    with maximum weight among all maximum-cardinality matchings.
+ * @param {string=} optWeight
+ *    The edge attribute that holds the numerical value used
+ *    as a weight. If null or not defined, then each edge has weight 1.
+ * @return {Map} mate
+ *    The matching is returned as a dictionary, mate, such that
+ *    mate[v] == w if node v is matched to node w. Unmatched nodes do not
+ *    occur as a key in mate.
+ */
+export function maxWeightMatching(G, maxcardinality=false, optWeight=undefined) {
+
+  // give each node an id in range [0, G.order-1]
+  const tr = new Map(zipIterator(G.nodesIter(), genRange(G.order( ))));
+  const edges = new Map();
+
+  if(!optWeight){
+    for (let [u, v] of G.edgesIter()) edges.set([tr.get(u), tr.get(v)], 1);
+  }
+  else {
+    // keep only max-weight edges for MultiGraph
+    for (let [u, v, d] of G.edgesIter(true)) {
+      const e = [tr.get(u), tr.get(v)];
+      const w = d[optWeight] != null ? d[optWeight] : 1;
+      if (!edges.has(e) || edges.get(e) < w) edges.set(e, w);
+    }
+  }
+
+  // compute maximum matching
+  const edgeslist = [for ([[u, v], w] of edges) [u, v, w]];
+  const mate = weightedBlossomEdmonds(false, false, false)(edgeslist, maxcardinality);
+
+  // translate ids back and return matching as a mapping
+  const trb = new Map([for ([key, value] of tr) [value, key]]);
+  return new Map( [for (i of genRange(mate.length)) if(mate[i] !== -1) [trb.get(i), trb.get(mate[i])]]);
+}
+
+
+// Adapted from http://jorisvr.nl/maximummatching.html
+// All credit for the implementation goes to Joris van Rantwijk [http://jorisvr.nl].
+
+// ** Original introduction below **
+
+// Weighted maximum matching in general graphs.
+
+// The algorithm is taken from "Efficient Algorithms for Finding Maximum
+// Matching in Graphs" by Zvi Galil, ACM Computing Surveys, 1986.
+// It is based on the "blossom" method for finding augmenting paths and
+// the "primal-dual" method for finding a matching of maximum weight, both
+// due to Jack Edmonds.
+// Some ideas came from "Implementation of algorithms for maximum matching
+// on non-bipartite graphs" by H.J. Gabow, Standford Ph.D. thesis, 1973.
+
+// A C program for maximum weight matching by Ed Rothberg was used extensively
+// to validate this new code.
+
+
+export function weightedBlossomEdmonds(debug, CHECK_OPTIMUM, CHECK_DELTA) {
+
+  // If assigned, DEBUG(str) is called with lots of debug messages.
+  var DEBUG = debug ? function(s){ console.log('DEBUG:', s); } : null;
```

Comment:
Do we need to keep the debug option of the original implementation or can we just drop it?